File: typescript/base/src/main/vsdk/toolkit/common/linealAlgebra/Vector3Dd.ts
export class Vector3Dd {
  public x: number;
  public y: number;
  public z: number;

  public constructor(inX = 0.0, inY = 0.0, inZ = 0.0) {
    this.x = inX;
    this.y = inY;
    this.z = inZ;
  }

  public distanceSquared(v2: Vector3Dd): number {
    const d = new Vector3Dd();
    d.x = v2.x - this.x;
    d.y = v2.y - this.y;
    d.z = v2.z - this.z;
    return d.x * d.x + d.y * d.y + d.z * d.z;
  }

  public dotProduct(b: Vector3Dd): number {
    return this.x * b.x + this.y * b.y + this.z * b.z;
  }

  public isNull(epsilon: number): boolean {
    return this.dotProduct(this) <= epsilon * epsilon;
  }

  public normalizeAndGivePreviousNorm(epsilon: number): number {
    let len = this.dotProduct(this);
    if (len <= 0.0) {
      return 0.0;
    }

    if (len <= 1.0 + epsilon && len >= 1.0 - epsilon) {
      len = 0.5 + 0.5 * len;
    }
    else {
      len = globalThis.Math.sqrt(len);
    }

    this.x /= len;
    this.y /= len;
    this.z /= len;

    return len;
  }

  public crossProduct(a: Vector3Dd, b: Vector3Dd): void {
    this.x = a.y * b.z - a.z * b.y;
    this.y = a.z * b.x - a.x * b.z;
    this.z = a.x * b.y - a.y * b.x;
  }

  public copy(source: Vector3Dd): void {
    this.x = source.x;
    this.y = source.y;
    this.z = source.z;
  }
}

File: cpp/base/src/main/vsdk/toolkit/common/linealAlgebra/Vector3Dd.cpp
/**
Routines for 3-d vectors
*/

#include "java/lang/Math.h"
#include "vsdk/toolkit/common/linealAlgebra/Vector3Dd.h"

Vector3Dd::Vector3Dd() {
    x = 0.0;
    y = 0.0;
    z = 0.0;
}

Vector3Dd::Vector3Dd(double inX, double inY, double inZ) {
    x = inX;
    y = inY;
    z = inZ;
}

Vector3Dd::~Vector3Dd() {
}

/**
Normalize a vector, return old magnitude
*/
double
Vector3Dd::normalizeAndGivePreviousNorm(double epsilon)
{
    double len = this->dotProduct(this); // Note: this starts being length ^ 2

    if ( len <= 0.0 ) {
        return 0.0;
    }

    if ( len <= 1.0 + epsilon && len >= 1.0 - epsilon) {
        // First order approximation
        len = 0.5 + 0.5 * len;
    } else {
        len = java::Math::sqrt(len);
    }

    x /= len;
    y /= len;
    z /= len;

    return len;
}

/**
Cross product of two vectors
result = a X b
*/
void
Vector3Dd::crossProduct(const Vector3Dd *a, const Vector3Dd *b)
{
    x = a->y * b->z - a->z * b->y;
    y = a->z * b->x - a->x * b->z;
    z = a->x * b->y - a->y * b->x;
}

/**
Returns squared distance between the two vectors
*/
double
Vector3Dd::distanceSquared(const Vector3Dd *v2) const {
    Vector3Dd d;

    d.x = v2->x - x;
    d.y = v2->y - y;
    d.z = v2->z - z;
    return d.x * d.x + d.y * d.y + d.z * d.z;
}
